go-controller/pkg/ovn/obj_retry.go

```diff
@@ -711,6 +797,103 @@ func (oc *Controller) updateResource(objectsToRetry *retryObjs, oldObj, newObj i
+		}
+		return nil
+	case factory.EgressNodeType:
+		oldNode := oldObj.(*kapi.Node)
+		newNode := newObj.(*kapi.Node)
+		// Initialize the allocator on every update,
+		// ovnkube-node/cloud-network-config-controller will make sure to
+		// annotate the node with the egressIPConfig, but that might have
+		// happened after we processed the ADD for that object, hence keep
+		// retrying for all UPDATEs.
+		if err := oc.initEgressIPAllocator(newNode); err != nil {
+			klog.V(5).Infof("Egress node initialization error: %v", err)
+		}
+		nodeEgressLabel := util.GetNodeEgressLabel()
+		oldLabels := oldNode.GetLabels()
+		newLabels := newNode.GetLabels()
+		_, oldHadEgressLabel := oldLabels[nodeEgressLabel]
+		_, newHasEgressLabel := newLabels[nodeEgressLabel]
+		// If the node is not labelled for egress assignment, just return
+		// directly, we don't really need to set the ready / reachable
+		// status on this node if the user doesn't care about using it.
+		if !oldHadEgressLabel && !newHasEgressLabel {
+			return nil
+		}
+		if oldHadEgressLabel && !newHasEgressLabel {
+			klog.Infof("Node: %s has been un-labelled, deleting it from egress assignment", newNode.Name)
+			oc.setNodeEgressAssignable(oldNode.Name, false)
+			if err := oc.deleteEgressNode(oldNode); err != nil {
+				return err
+			}
+			return nil
+		}
+		isOldReady := oc.isEgressNodeReady(oldNode)
+		isNewReady := oc.isEgressNodeReady(newNode)
+		isNewReachable := oc.isEgressNodeReachable(newNode)
+		oc.setNodeEgressReady(newNode.Name, isNewReady)
+		oc.setNodeEgressReachable(newNode.Name, isNewReachable)
+		if !oldHadEgressLabel && newHasEgressLabel {
+			klog.Infof("Node: %s has been labelled, adding it for egress assignment", newNode.Name)
+			oc.setNodeEgressAssignable(newNode.Name, true)
+			if isNewReady && isNewReachable {
+				if err := oc.addEgressNode(newNode); err != nil {
+					return err
+				}
+			} else {
+				klog.Warningf("Node: %s has been labelled, but node is not ready and reachable, cannot use it for egress assignment", newNode.Name)
+			}
+			return nil
+		}
+		if isOldReady == isNewReady {
+			return nil
+		}
+		if !isNewReady {
+			klog.Warningf("Node: %s is not ready, deleting it from egress assignment", newNode.Name)
+			if err := oc.deleteEgressNode(newNode); err != nil {
+				return err
+			}
+		} else if isNewReady && isNewReachable {
+			klog.Infof("Node: %s is ready and reachable, adding it for egress assignment", newNode.Name)
+			if err := oc.addEgressNode(newNode); err != nil {
+				return err
+			}
+		}
+		return nil
+	case factory.CloudPrivateIPConfigType:
+		oldCloudPrivateIPConfig := oldObj.(*ocpcloudnetworkapi.CloudPrivateIPConfig)
+		newCloudPrivateIPConfig := newObj.(*ocpcloudnetworkapi.CloudPrivateIPConfig)
+		if err := oc.reconcileCloudPrivateIPConfig(oldCloudPrivateIPConfig, newCloudPrivateIPConfig); err != nil {
+			return fmt.Errorf("unable to update CloudPrivateIPConfig: %s, err: %v", newCloudPrivateIPConfig.Name, err)
+		}
+		return nil
```

Comment:
Hmm... the update code for egress nodes is quite complex.

If ever `updateResource` fails, its caller (`watchResource`) will just set `newObj` in the corresponding retry entry and will return. So all the logic in `updateResource` for egressIPNodes will be ignored and we'll just try to add `newObj` at the next iteration of  `iterateRetryResources`. This is bad, right?

```
				if hasUpdateFunc {
					klog.Infof("Updating %s %s", objectsToRetry.oType, newKey)
					// if this resource type has an update func, just call the update function
					if err := oc.updateResource(objectsToRetry, old, newer); err != nil {
						klog.Errorf("Failed to update %s, old=%s, new=%s, error: %v",
							objectsToRetry.oType, oldKey, newKey, err)
						oc.recordErrorEvent(objectsToRetry.oType, newer, err)
						objectsToRetry.initRetryObjWithAdd(newer, newKey)
						objectsToRetry.unSkipRetryObj(newKey)
						return
					}

```

Can we compensate by setting both `oldObj` and `newObj`, so that in `iterateRetryResources` we first delete the old instance and then we try adding the new one?
Or actually, in `iterateRetryResources`, why don't we call `updateResource` if both old and new obj are set?

@trozet what do you think?

Reply:
I agree not all resouces can handle update via delete old then add new gracefully like for eIP we should allow iterate to handle update if the resource needs I will spin new api and for only for egressNode to be used by the iterate logic

Reply:
I think it might lead to some unexpected behavior. The reason is that oldObj in this case can be one of two things:
1. either its an old version of this instance, which we could just assume update method for
2. or its actually a version of an older instance, in which case we dont know if the controllers have accounted for such a scenario (probably not).

In scenario 2 the controllers would have to be capable of "updating" from one instance to another, which I can already see is going to be in uncharted territory. Like the pod we delete the old instance with a different IP and annotations, and create a new instance. We never thought about cases like preserving the old instance annotations and copying them to the new.

It sounds like what we really want here is "retry update". Let me take a closer look at the egressIP code and see if there is any obvious solution. I thought Alex made egress IP more level driven so we would just be able to call add and not worry about update so this is a bit surprising. Perhaps one solution is to store the update info in the config property of the retry struct and then egress IP can leverage that to know its an update or something.

Reply:
Alternatively to solve number 2 you could check if the UID is different between the two objects. If it is then you know you need to delete the old and add the new, otherwise you can do update? wdyt?